neo/SmartContract/NativeContracts/GasToken.cs

```diff
@@ -0,0 +1,170 @@
+                return false;
+            string operation = engine.CurrentContext.EvaluationStack.Pop().GetString();
+            VMArray args = (VMArray)engine.CurrentContext.EvaluationStack.Pop();
+            StackItem result;
+            switch (operation)
+            {
+                case "supportedStandards":
+                    result = GasToken_SupportedStandards.Select(p => (StackItem)p).ToList();
+                    break;
+                case "name":
+                    result = GasToken_Name;
+                    break;
+                case "symbol":
+                    result = GasToken_Symbol;
+                    break;
+                case "decimals":
+                    result = GasToken_Decimals;
+                    break;
+                case "totalSupply":
+                    result = GasToken_TotalSupply();
+                    break;
+                case "balanceOf":
+                    result = GasToken_BalanceOf(engine, args[0].GetByteArray());
+                    break;
+                case "transfer":
+                    result = GasToken_Transfer(engine, args[0].GetByteArray(), args[1].GetByteArray(), args[2].GetBigInteger());
+                    break;
+                default:
+                    return false;
+            }
+            engine.CurrentContext.EvaluationStack.Push(result);
+            return true;
+        }
+
+        private BigInteger GasToken_TotalSupply()
+        {
+            StorageItem storage = Snapshot.Storages.TryGet(new StorageKey
+            {
+                ScriptHash = Blockchain.GasToken.ScriptHash,
+                Key = Encoding.ASCII.GetBytes("totalSupply")
+            });
+            if (storage is null) return BigInteger.Zero;
+            return new BigInteger(storage.Value);
+        }
+
+        private BigInteger GasToken_BalanceOf(ExecutionEngine engine, byte[] account)
+        {
+            if (account.Length != 20) throw new ArgumentException();
+            StorageItem storage = Snapshot.Storages.TryGet(new StorageKey
+            {
+                ScriptHash = Blockchain.GasToken.ScriptHash,
+                Key = account
+            });
+            if (storage is null) return BigInteger.Zero;
+            Struct state = (Struct)storage.Value.DeserializeStackItem(engine.MaxArraySize);
+            return state[0].GetBigInteger();
+        }
+
+        private bool GasToken_Transfer(ExecutionEngine engine, byte[] from, byte[] to, BigInteger amount)
+        {
+            if (Trigger != TriggerType.Application) throw new InvalidOperationException();
+            UInt160 hash_from = new UInt160(from);
+            UInt160 hash_to = new UInt160(to);
+            if (amount.Sign < 0) throw new ArgumentOutOfRangeException(nameof(amount));
+            if (!hash_from.Equals(new UInt160(engine.CurrentContext.CallingScriptHash)) && !CheckWitness(engine, new UInt160(from)))
```

Comment:
In wich case we need this condition? `!hash_from.Equals(new UInt160(engine.CurrentContext.CallingScriptHash)) && `

Reply:
It seems this is for the case where another contract calls this contract.

Reply:
It is according to NEP-5 standard: https://github.com/neo-project/proposals/blob/master/nep-5.mediawiki#transfer

> The function SHOULD check whether the `from` address equals the caller contract hash. If so, the transfer SHOULD be processed; If not, the function SHOULD use the SYSCALL `Neo.Runtime.CheckWitness` to verify the transfer.